tiledb/cloud/cloudarray.py

```diff
@@ -144,3 +163,32 @@ def apply(self, func, subarray, attrs=None, layout=None):
         except GenApiException as exc:
             raise tiledb_cloud_error.check_udf_exc(exc) from None
         return cloudpickle.loads(res)
+
+    def _calculate_ranges(self, subarray):
+        ranges = list()
+        if isinstance(subarray, MultiIndexRanges):
+            for dim_idx, dim_rng in enumerate(subarray.ranges):
+                dim = self.schema.domain.dim(dim_idx)
+                for idx, rng in enumerate(dim_rng):
+                    start = build_dimension_coordinate(dim.dtype, rng[0])
+                    end = build_dimension_coordinate(dim.dtype, rng[1] + 1)
```

Comment:
Should we be adding a +1 here?

Reply:
nope, thanks -- left over from testing on prod when it was endpoint-exclusive

Reply:
adding some tests for apply now